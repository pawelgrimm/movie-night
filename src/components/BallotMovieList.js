import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import { useParams } from "react-router-dom";
import MovieListItem from "./MovieListItem";
import { ThumbsUp, ThumbsDown } from "react-feather";
import { openModal } from "../actions/videoModal";
import { getMoviesOnBallot } from "../firebase/api";
import Loader from "./Loader";
import { hydrateMovies } from "../selectors/movie";
import { fetchMovie } from "../actions/movie";

const mapStateToProps = ({ movies }) => ({
  savedMovies: movies,
});

const mapDispatchToProps = (dispatch) => ({
  openModal: (videos) => dispatch(openModal(videos)),
  fetchMovie: (id) => dispatch(fetchMovie(id)),
});

export const BallotMovieList = ({ savedMovies, openModal, fetchMovie }) => {
  const [loaded, setLoaded] = useState(false);
  const [movies, setMovies] = useState([]);
  const [movieIds, setMovieIds] = useState([]);
  const { id } = useParams();

  useEffect(() => {
    getMoviesOnBallot(id).then((ref) => {
      const movieIds = ref.val();
      movieIds.forEach((id) => fetchMovie(id));
      setMovieIds(movieIds);
      setMovies(hydrateMovies(savedMovies, movieIds));
      setLoaded(true);
    });
  }, []);

  useEffect(() => {
    if (loaded) {
      setMovies(hydrateMovies(savedMovies, movieIds));
    }
  }, [savedMovies]);

  if (!loaded) {
    return <Loader />;
  }

  return (
    <div className="list-body">
      {movies.map((movie) => {
        const id = movie.info.id;
        return (
          <MovieListItem
            key={id}
            isLoading={!movie.isSaved}
            movie={movie.info}
            collapsedButtons={collapsedButtons(id)}
            expandedButtons={expandedButtons(
              id,
              movie.info.trailers,
              openModal
            )}
          />
        );
      })}
    </div>
  );
};

const collapsedButtons = (movieId) => makeButtons(movieId, true);

const expandedButtons = (movieId, videos, openModal) => [
  videos && videos.length > 0 && (
    <button
      className="button"
      onClick={(event) => {
        event.stopPropagation();
        openModal(videos);
      }}
    >
      Watch Trailer
    </button>
  ),
  ...makeButtons(movieId, false),
];

const makeButtons = (movieId, collapsed = true) => [
  <button
    className="button"
    onClick={(event) => {
      event.stopPropagation();
      console.log("thumbs up clicked for", movieId);
    }}
  >
    <ThumbsUp />
  </button>,
  <button
    className="button"
    onClick={(event) => {
      event.stopPropagation();
      console.log("thumbs up clicked for", movieId);
    }}
  >
    <ThumbsDown />
  </button>,
];

export default connect(mapStateToProps, mapDispatchToProps)(BallotMovieList);
