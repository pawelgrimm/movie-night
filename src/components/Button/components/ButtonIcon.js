import React from "react";

const ButtonIcon = ({ children, ...rest }) => {
  return <>{children}</>;
};

export default ButtonIcon;
