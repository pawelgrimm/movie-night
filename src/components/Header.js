import React from "react";
import { Link } from "react-router-dom";

export const Header = () => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link
          className="header__title"
          to="/dashboard"
        >
          <h1>Movie Night</h1>
        </Link>
      </div>
    </div>
  </header>
);

export default Header;
