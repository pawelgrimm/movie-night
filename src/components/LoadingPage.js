import React from "react";
import Loader from "./Loader";

const LoadingPage = () => (
  <div className="loader">
    <Loader />
  </div>
);

export default LoadingPage;
