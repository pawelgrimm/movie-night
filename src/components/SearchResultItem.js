import React, { useContext, useEffect, useReducer, useState } from "react";
import ReactImageFallback from "react-image-fallback";
import { addMovie, removeMovie } from "../actions/movie";
import MovieContext from "../context/movie-context";
import MovieImage from "./MovieImage";

const SearchResultItem = ({ movie, resetSearch }) => {
  const { movies: savedMovies, dispatch } = useContext(MovieContext);
  const [button, setButton] = useState("+");
  const [buttonClass, setButtonClass] = useState("");

  useEffect(() => {
    if (savedMovies.find((savedMovie) => savedMovie.id === movie.id)) {
      toggleButtonState();
    }
  }, []);

  const toggleButtonState = () => {
    setButton(button === "+" ? "-" : "+");
    setButtonClass(buttonClass ? "" : "button--secondary");
  };

  const onClick = () => {
    toggleButtonState();
    if (button === "+") {
      dispatch(addMovie(movie));
      resetSearch();
    } else {
      dispatch(removeMovie(movie.id));
    }
  };
  return (
    <div className="search-result-item">
      <MovieImage
        title={movie.title}
        imageWidth={68}
        filePath={movie.poster_path}
        type="poster"
        className="search-result-item__poster"
      />
      <div className="search-result-item__info">
        <h3>{movie.title}</h3>
        <span>{movie.release_date}</span>
        <p>{movie.overview}</p>
      </div>
      <div className="search-result-item__button">
        <button className={"button " + buttonClass} onClick={onClick}>
          {button}
        </button>
      </div>
    </div>
  );
};

export default SearchResultItem;
