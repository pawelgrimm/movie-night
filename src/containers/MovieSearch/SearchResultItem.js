import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import MovieImage from "../../components/MovieImage/MovieImage";
import { useSearchContext } from "./search-context";
import { addAndSaveMovie, removeMovie } from "../../services/ballot/actions";
import { formatReleaseYear } from "../../services/theMovieDb/theMovieDb";

const mapStateToProps = ({ ballot }) => ({
  movies: ballot.movies,
});

const mapDispatchToProps = (dispatch) => ({
  addMovie: (movie) => {
    dispatch(addAndSaveMovie(movie.id, movie));
  },
  removeMovie: (id) => dispatch(removeMovie(id)),
});

export const SearchResultItem = ({ movie, movies, addMovie, removeMovie }) => {
  const [button, setButton] = useState("+");
  const [buttonClass, setButtonClass] = useState("");
  const { resetSearch } = useSearchContext();

  useEffect(() => {
    if (movies.find((id) => id === movie.id)) {
      toggleButtonState();
    }
  }, []);

  const toggleButtonState = () => {
    setButton(button === "+" ? "-" : "+");
    setButtonClass(buttonClass ? "" : "button--secondary");
  };

  const onClick = () => {
    toggleButtonState();
    if (button === "+") {
      addMovie(movie);
    } else {
      removeMovie(movie.id);
    }
    resetSearch();
  };
  return (
    <div className="search-result-item" tabIndex={0}>
      <MovieImage
        title={movie.title}
        imageWidth={68}
        filePath={movie.poster_path}
        type="poster"
        className="search-result-item__poster"
      />
      <div className="search-result-item__info">
        <h3>{movie.title}</h3>
        <span>{formatReleaseYear(movie.release_date)}</span>
        <p>{movie.overview}</p>
      </div>
      <div className="search-result-item__button">
        <button className={"button " + buttonClass} onClick={onClick}>
          {button}
        </button>
      </div>
    </div>
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchResultItem);
