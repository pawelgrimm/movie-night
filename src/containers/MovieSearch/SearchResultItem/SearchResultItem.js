import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
import MovieImage from "../../../components/MovieImage/MovieImage";
import { useSearchContext } from "../search-context";
import { addAndSaveMovie, removeMovie } from "../../../services/ballot/actions";
import { formatReleaseYear } from "../../../utils/theMovieDb";
import style from "./search-result-item.module.scss";
import Button from "../../../components/Button/Button";

const mapStateToProps = ({ ballot }) => ({
  movies: ballot.movies,
});

const mapDispatchToProps = (dispatch) => ({
  addMovie: (movie) => {
    dispatch(addAndSaveMovie(movie.id, movie));
  },
  removeMovie: (id) => dispatch(removeMovie(id)),
});

export const SearchResultItem = ({ movie, movies, addMovie, removeMovie }) => {
  const [button, setButton] = useState("+");
  const [isSecondary, setIsSecondary] = useState(false);
  const { resetSearch } = useSearchContext();

  useEffect(() => {
    if (movies.find((id) => id === movie.id)) {
      toggleButtonState();
    }
  }, []);

  const toggleButtonState = () => {
    setButton(button === "+" ? "-" : "+");
    setIsSecondary((prev) => !prev);
  };

  const onClick = () => {
    toggleButtonState();
    if (button === "+") {
      addMovie(movie);
    } else {
      removeMovie(movie.id);
    }
    resetSearch();
  };
  return (
    <div className={style.resultItem} tabIndex={0}>
      <MovieImage
        title={movie.title}
        imageWidth={68}
        filePath={movie.poster_path}
        type="poster"
        className={style.poster}
      />
      <div className={style.info}>
        <h3>{movie.title}</h3>
        <span>{formatReleaseYear(movie.release_date)}</span>
        <p>{movie.overview}</p>
      </div>
      <div className={style.buttons}>
        <Button type={isSecondary} className={style.button} onClick={onClick}>
          {button}
        </Button>
      </div>
    </div>
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchResultItem);
