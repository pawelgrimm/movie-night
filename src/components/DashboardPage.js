import React, {useReducer} from "react";
import MovieSearch from "./MovieSearch";
import searchReducer from "../reducers/search";
import SearchContext from "../context/search-context";

const DashboardPage = () => {
    const [ searchResults, dispatch ] = useReducer(searchReducer, []);
    const setBackground = (imageUrl) => {

    }
    return (
        <SearchContext.Provider value={{searchResults, dispatch}}>
            <div className="content-container">
                <MovieSearch />
            </div>
        </SearchContext.Provider>
    );
}

export default DashboardPage;
