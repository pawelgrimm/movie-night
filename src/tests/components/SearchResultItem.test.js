import { expect } from "@jest/globals";
import { shallow } from "enzyme";
import React from "react";
import { SearchResultItem } from "../../components/SearchResultItem";
import { movies } from "../fixtures/fixtures";

test("should render search result item correctly", () => {
  const movie = movies[2];
  const wrapper = shallow(
    <SearchResultItem
      movies={[]}
      addMovie={() => {}}
      removeMovie={() => {}}
      movie={movie}
      resetSearch={() => {}}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
