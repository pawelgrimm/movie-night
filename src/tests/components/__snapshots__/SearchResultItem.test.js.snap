// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render search result item correctly 1`] = `
<div
  className="search-result-item"
>
  <MovieImage
    className="search-result-item__poster"
    filePath="/gycdE1ARByGQcK4fYR2mgpU6OO.jpg"
    imageWidth={68}
    title="Before Sunset"
    type="poster"
  />
  <div
    className="search-result-item__info"
  >
    <h3>
      Before Sunset
    </h3>
    <span>
      2004
    </span>
    <p>
      Nine years ago two strangers met by chance and spent a night in Vienna that ended before sunrise. They are about to meet for the first time since. Now they have one afternoon to find out if they belong together.
    </p>
  </div>
  <div
    className="search-result-item__button"
  >
    <button
      className="button "
      onClick={[Function]}
    >
      +
    </button>
  </div>
</div>
`;
